server/grpc_test.go

```diff
@@ -410,6 +412,95 @@ func TestGrpcAcRequestInlinedBlobs(t *testing.T) {
 	}
 }
 
+func TestGrpcByteStreamDeadline(t *testing.T) {
+	testBlobSize := int64(16)
+	testBlob, testBlobHash := testutils.RandomDataAndHash(testBlobSize)
+	testBlobDigest := pb.Digest{
+		Hash:      testBlobHash,
+		SizeBytes: int64(len(testBlob)),
+	}
+
+	instance := "deadlineExpired"
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond*10)
+	defer cancel()
+
+	bswc, err := bsClient.Write(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	resourceName := fmt.Sprintf(
+		"%s/uploads/%s/blobs/%s/%d/deadline/metadata/here",
+		instance,
+		uuid.New().String(),
+		testBlobDigest.Hash,
+		len(testBlob),
+	)
+
+	for i := 0; i < len(testBlob); i++ {
+		bswReq := bytestream.WriteRequest{
+			ResourceName: resourceName,
+			FinishWrite:  false,
+			Data:         testBlob[i : i+1],
+			WriteOffset:  int64(i),
+		}
+
+		err := bswc.Send(&bswReq)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		time.Sleep(time.Millisecond)
+	}
+
+	_, err = bswc.CloseAndRecv()
+
+	statusError, ok := status.FromError(err)
+	if !ok {
+		t.Fatalf("expected rpc error code, got %v\n", err)
+	}
+
+	if code := statusError.Code(); code != codes.DeadlineExceeded {
+		t.Fatalf("expected codes.DeadlineExceeded, got %s\n", code.String())
+	}
+
+	ctx, cancel = context.WithTimeout(context.Background(), time.Millisecond*500)
+	defer cancel()
+
+	bswc, err = bsClient.Write(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	bswReq := bytestream.WriteRequest{
+		ResourceName: resourceName,
+		FinishWrite:  false,
+		Data:         testBlob,
+	}
+	err = bswc.Send(&bswReq)
+	if err != nil {
+		t.Errorf("send error: %v\n", err)
+	}
+
+	_, err = bswc.CloseAndRecv()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, sz, err := diskCache.Get(cache.CAS, testBlobHash)
+	if err != nil {
+		t.Errorf("get error: %v\n", err)
```

Comment:
I would make this `t.Fatalf(...)` also, since it is usually expected that if a function in go returns multiple values including a non-nil error, that the other return values should not be used (and so we should not check `sz` below if we get an error here).

Reply:
done